File: Lab 11/11-2.py
class Solution(object):
    def checkValidString(self, s):
    
        def dfs(index, openCount):
            if index == len(s):
                return openCount == 0
            
            if openCount < 0:
                return False
            
            char = s[index]
            
            if char == '(':
                return dfs(index + 1, openCount + 1)
            elif char == ')':
                return dfs(index + 1, openCount - 1)
            elif char == '*':
                return (
                    dfs(index + 1, openCount + 1) or 
                    dfs(index + 1, openCount - 1) or  
                    dfs(index + 1, openCount)         
                )
        
        return dfs(0, 0)

File: Lab 11/11-2.java
class Solution {
    public boolean checkValidString(String s) {
        int openLow = 0, openHigh = 0;

        for (char c : s.toCharArray()) {
            if (c == '(') {
                openLow++;
                openHigh++;
            } else if (c == ')') {
                openLow = Math.max(0, openLow - 1);
                openHigh--;
            } else { // c == '*'
                openLow = Math.max(0, openLow - 1);
                openHigh++;
            }

            if (openHigh < 0) {
                return false;
            }
        }

        return openLow == 0;
    }
}
